diff --git a/packages/WAPPushManager/Android.mk b/packages/WAPPushManager/Android.mk
index 91526dd1..b1ced73e 100644
--- a/packages/WAPPushManager/Android.mk
+++ b/packages/WAPPushManager/Android.mk
@@ -12,7 +12,7 @@ LOCAL_PACKAGE_NAME := WAPPushManager
 LOCAL_PRIVATE_PLATFORM_APIS := true
 
 LOCAL_JAVA_LIBRARIES += telephony-common
-LOCAL_STATIC_JAVA_LIBRARIES += android-common
+# android-common removed: not present in trimmed Halium tree (not needed for UT)
 
 LOCAL_PROGUARD_FLAG_FILES := proguard.flags
 
